### PR TITLE
Adding custom op ConvTransposeWithDynamicPads.

### PR DESCRIPTION
**Pull request by duli2012**
This custom op intend to fix the below issue:
https://github.com/onnx/onnx/issues/1772
Once the ONNX spec changes accordingly, this op will replace the ONNX ConvTranspose op.

Since this op will become onnx op, the doc and shape inference will be added ONNX side, which is omitted from kernel implementation for now. This also keeps binary size impact minimal.

### CODE REVIEW

File: cmake/onnxruntime_providers.cmake [hunk 1]
```diff
@@ -6,11 +6,13 @@ file(GLOB_RECURSE onnxruntime_providers_srcs CONFIGURE_DEPENDS
   "${ONNXRUNTIME_ROOT}/core/providers/cpu/*.cc"
 )
 
-file(GLOB_RECURSE onnxruntime_contrib_ops_srcs CONFIGURE_DEPENDS
```
Comment (pranavsharma) [inline cmake/onnxruntime_providers.cmake]: Can we keep this? It was added by someone outside our team and he seemed to have good reasons to use this.

File: onnxruntime/contrib_ops/contrib_kernels_cuda.cc [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (pranavsharma) [inline onnxruntime/contrib_ops/contrib_kernels_cuda.cc]: Can this file be present in the cuda subdirectory? For our regular ops, we keep the file that registers all the ops (cuda/cuda_execution_provider.cc/h) in the same directory where the kernels are present. Would be good to remain consistent.

File: onnxruntime/core/providers/cpu/nn/conv_transpose.cc [hunk 3]
```diff
@@ -129,7 +130,14 @@ Status ConvTransposeBase::PrepareForCompute(OpKernelContext* context, bool has_b
   if (output_padding.empty()) {
     output_padding.resize(kernel_shape.size(), 0);
   }
-  std::vector<int64_t> pads(pads_);
+  std::vector<int64_t> pads;
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/nn/conv_transpose.cc]: can you `reserve()` the size of `pads`?

File: cmake/onnxruntime_providers.cmake (hunk 1 above)
Comment (duli2012) [reply cmake/onnxruntime_providers.cmake]: done

File: onnxruntime/contrib_ops/contrib_kernels_cuda.cc (hunk 2 above)
Comment (duli2012) [reply onnxruntime/contrib_ops/contrib_kernels_cuda.cc]: moved

File: onnxruntime/core/providers/cpu/nn/conv_transpose.cc (hunk 3 above)
Comment (duli2012) [reply onnxruntime/core/providers/cpu/nn/conv_transpose.cc]: Added.

File: onnxruntime/core/providers/cpu/nn/conv_transpose.cc [hunk 4]
```diff
@@ -53,30 +53,31 @@ inline void ComputeTransposePadAndOutputShape(
...
 }
 
-Status ConvTransposeBase::PrepareForCompute(OpKernelContext* context, bool has_bias, ConvTransposeBase::Prepare& p) const {
-  const auto* X = context->Input<Tensor>(0);
-  const auto* F = context->Input<Tensor>(1);
-  const Tensor* B = has_bias ? context->Input<Tensor>(2) : nullptr;
+Status ConvTransposeBase::PrepareForCompute(OpKernelContext* context, bool has_bias, ConvTransposeBase::Prepare& p, bool dynamic_padding) const {
+  const Tensor* X = context->Input<Tensor>(0);
+  const Tensor* F = context->Input<Tensor>(1);
+  const Tensor* Pads = dynamic_padding ? context->Input<Tensor>(2) : nullptr;
```
Comment (zhangxiang1993) [inline onnxruntime/core/providers/cpu/nn/conv_transpose.cc]: In the schema, Pads is optional. But it seems that the kernel assume pads exist.

**Pull request merged**